assets/model_monitoring/components/src/action_analyzer/action_analyzer_relevance_action_detector/relevance_action_detector.py

```diff
@@ -0,0 +1,606 @@
+            response["response_time_sec"] = time_taken
+            rating = float(response["samples"][0])
+            ratings.append(rating)
+        except Exception as e:  # noqa: B902
+            response["finish_reason"] = ["error"]
+            response["error"] = [str(e)]
+            raise e
+    return ratings[0]
+
+
+def _post_process_results(output):
+    parsed_score_response = re.findall(r'\d+', output.split("# Result")[-1].strip())
+    if len(parsed_score_response) > 0:
+        score = float(parsed_score_response[0].replace("'", "").strip())
+    else:
+        # Result of score is not found in the output string
+        score = DEFAULT_RETRIEVAL_SCORE
+        print("Not able to parse the score from the output string, setting score to nan")
+    return score
+
+
+def _query_retrieval_score(
+    turns: Tuple[str, str, str],
+    session: requests.Session,
+    endpoint_url: str,
+    token_manager: _APITokenManager,
+    model: str,
+    temperature: float,
+    top_p: float,
+    num_samples: int,
+    frequency_penalty: float,
+    presence_penalty: float,
+    max_tokens=3000,
+    stop: str = None
+) -> int:
+
+    turns_list = [turns]
+    prompts = [RELEVANCE_TEMPLATE.replace("{{ query }}", turn[0]).replace("{{ history }}", "").replace("{{ FullBody }}", turn[2]) for turn in turns_list]  # noqa: E501
+
+    ratings = []
+    for prompt in prompts:
+        request_data = {
+            "model": model,
+            "temperature": temperature,
+            "top_p": top_p,
+            "n": num_samples,
+            "max_tokens": max_tokens,
+            "frequency_penalty": frequency_penalty,
+            "presence_penalty": presence_penalty,
+            "messages": [
+                {"role": "user", "content": prompt}
+            ]
+        }
+
+        if stop:
+            request_data["stop"] = stop
+
+        response = {}
+        try:
+            response, time_taken = _request_api(
+                session=session,
+                endpoint_url=endpoint_url,
+                token_manager=token_manager,
+                **request_data,
+            )
+            # Append time taken to the line
+            response["response_time_sec"] = time_taken
+            rating = _post_process_results(response["samples"][0])
+            ratings.append(rating)
+        except Exception as e:  # noqa: B902
+            response["finish_reason"] = ["error"]
+            response["error"] = [str(e)]
+            raise e
+    return ratings[0]
+
+
+def get_llm_score(row,
+                  workspace_connection_arm_id,
+                  model_deployment_name,
+                  api_call_retry_max_count,
+                  api_call_retry_backoff_factor,
+                  request_args,
+                  score_name):
+    """Get llm score."""
+    question = row[MODIFIED_PROMPT_COLUMN]
+    answer = row[COMPLETION_COLUMN]
+    text = row[RETRIEVAL_DOC_COLUMN]
+
+    token_manager = _WorkspaceConnectionTokenManager(
```

Comment:
this is created multiple times in different functions, shall we create it once and pass it along?